<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="msCentipede : A hierarchical multiscale model for inferring transcription factor binding from chromatin accessibility data.">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>msCentipede</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/rajanil/msCentipede">View on GitHub</a>

          <h1 id="project_title">msCentipede</h1>
          <h2 id="project_tagline">A hierarchical multiscale model for inferring transcription factor binding from chromatin accessibility data.</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/rajanil/msCentipede/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/rajanil/msCentipede/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h1>
<a id="mscentipede" class="anchor" href="#mscentipede" aria-hidden="true"><span class="octicon octicon-link"></span></a>msCentipede</h1>

<p><strong>msCentipede</strong> is an algorithm for accurately inferring transcription factor binding sites using chromatin
accessibility data (Dnase-seq, ATAC-seq) and is written in Python2.x and Cython. 
The <a href="http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0138030">hierarchical multiscale model underlying msCentipede</a> identifies factor-bound genomic sites
by using patterns in DNA cleavage resulting from the action of nucleases in open chromatin regions 
(regions typically bound by transcription factors). msCentipede, 
a generalization of the <a href="http://centipede.uchicago.edu">CENTIPEDE</a> model, accounts for 
heterogeneity in the DNA cleavage patterns around sites bound by transcription factors.</p>

<p>This code repository contains set of scripts to load the data and run the algorithm. The current document summarizes 
how to download and setup this software package and provides instructions on how to run the software
on a test dataset of motif instances and some publicly available DNase-seq data.</p>

<h2>
<a id="dependencies" class="anchor" href="#dependencies" aria-hidden="true"><span class="octicon octicon-link"></span></a>Dependencies</h2>

<p>msCentipede depends on </p>

<ul>
<li><a href="http://www.numpy.org/">Numpy</a></li>
<li><a href="http://www.scipy.org/">Scipy</a></li>
<li><a href="http://cython.org/">Cython</a></li>
<li><a href="http://www.cvxopt.org/">Cvxopt</a></li>
<li><a href="https://github.com/pysam-developers/pysam">Pysam</a></li>
</ul>

<p>A number of python distributions already have the first two modules packaged in them. It is also
straightforward to install all these dependencies 
 (1) using package managers for MACOSX and several Linux distributions,
 (2) from platform-specific binary packages, and
 (3) directly from source</p>

<h2>
<a id="getting-the-source-code" class="anchor" href="#getting-the-source-code" aria-hidden="true"><span class="octicon octicon-link"></span></a>Getting the source code</h2>

<p>To obtain the source code from github, let us assume you want to clone this repo into a
directory named <code>proj</code>:</p>

<pre><code>mkdir ~/proj
cd ~/proj
git clone https://github.com/rajanil/msCentipede
</code></pre>

<p>To retrieve the latest code updates, you can do the following:</p>

<pre><code>cd ~/proj/msCentipede
git fetch
git merge origin/master
</code></pre>

<p>Since the algorithm is written in Cython, msCentipede will have to be compiled into a shared object in the following way, before it can be executed.</p>

<pre><code>python setup.py build_ext --inplace
</code></pre>

<p>This step will generate a number of warning messages that can be ignored. If there are no error messages, the compilation step should produce two files <code>mscentipede.c</code> and <code>mscentipede.so</code>.</p>

<h2>
<a id="executing-the-code" class="anchor" href="#executing-the-code" aria-hidden="true"><span class="octicon octicon-link"></span></a>Executing the code</h2>

<p>The script you will need to execute is <code>call_binding.py</code>. To see command-line 
options that need to be passed to the script, you can do the following:</p>

<pre><code>$ python call_binding.py

runs msCentipede, to infer transcription factor binding, given a set of motif
instances and chromatin accessibility data

positional arguments:
  motif_file            name of a gzipped text file containing positional
                        information and other attributes for motif instances
                        of a transcription factor. columns of the file should
                        be as follows. Chromosome Start End Strand PWM_Score
                        [Attribute_1 Attribute_2 ...]. additional attributes
                        are optional.
  bam_files             whitespace separated list of bam files from a
                        chromatin accessibility assay

optional arguments:
  -h, --help            show this help message and exit
  --task {learn,infer}  specify whether to learn model parameters or infer
                        factor binding (default: learn)
  --protocol {ATAC_seq,DNase_seq}
                        specifies the chromatin accessibility protocol
                        (default:DNase_seq)
  --model {msCentipede,msCentipede_flexbg,msCentipede_flexbgmean}
                        models differ in how they capture background rate of
                        enzyme cleavage (default:msCentipede)
  --restarts RESTARTS   number of re-runs of the algorithm (default: 1)
  --mintol MINTOL       convergence criterion for change in per-site marginal
                        likelihood (default: 1e-6)
  --model_file MODEL_FILE
                        file name to store the model parameters
  --posterior_file POSTERIOR_FILE
                        file name to store the posterior odds ratio, and
                        likelihood ratios for each model component, at each
                        motif.
  --log_file LOG_FILE   file name to store some statistics of the EM algorithm
  --window WINDOW       size of window around the motif instance, where
                        chromatin accessibility profile is used for inferring
                        transcription factor binding. (default: 128)
  --batch BATCH         maximum number of motif instances used for learning
                        model parameters. this is also the number of motif
                        instances on which inference is performed at a time.
                        (default: 10000)
  --bam_file_genomicdna BAM_FILE_GENOMICDNA
                        bam file from a chromatin accessibility assay on
                        genomic DNA
  --seed SEED           set seed for random initialization of parameters
</code></pre>

<p>We will now describe in detail how to use this software using an example dataset of CTCF motif instances on chromosome 10 in hg19 coordinates is provided in <code>test/CTCF_chr10_motifs.txt.gz</code>. DNase-seq data for the GM12878 cell line (bam and bai files) can be downloaded from ENCODE to <code>test/</code> . In the following instructions, we assume the data files are named <a href="http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeUwDnase/wgEncodeUwDnaseGm12878AlnRep1.bam">Gm12878_Rep1.bam</a> and <a href="http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeUwDnase/wgEncodeUwDnaseGm12878AlnRep2.bam">Gm12878_Rep2.bam</a>.</p>

<p>The software is designed to run in two separate steps. In the first step, optimal values for the model parameters are estimated using a subset of all motif instances. In the second step, posterior probability of factor binding is inferred for all motif instances. Since accurate estimates of model parameters can be obtained using 5000-10000 motif instances, this enables efficient inference for those transcription factors that have orders of magnitude more motif instances genomewide. If more motif instances are available in the file than the value of the flag <code>--batch</code>, then <code>batch</code> number of motif instances that have the highest PWM score are used in learning model parameters.</p>

<h3>
<a id="key-inputs" class="anchor" href="#key-inputs" aria-hidden="true"><span class="octicon octicon-link"></span></a>Key Inputs</h3>

<p>The key inputs that need to be passed to this script are </p>

<ul>
<li>  a path to the file containing the list of motif instances</li>
<li>
<p>the bam files (sorted and indexed) containing sequencing reads from a chromatin accessibility assay (DNase-seq or ATAC-seq). </p>

<p><em>Note: these inputs are positional arguments and the files must be specified in the correct order (as shown above).</em> </p>
</li>
</ul>

<p>The gzipped file of motif instances should have the following format.</p>

<pre><code>Chr   Start     Stop      Strand  PwmScore
chr10 3944439   3944456   +       15.21570492
chr10 15627426  15627443  -       20.39377594
</code></pre>

<p>In the above format, positions are 0-based. <em>Start</em> corresponds to the first base of the core motif for <em>+</em> strand motif instances and the last base of the core motif for <em>-</em> strand motif instances.</p>

<p>When multiple library / sample replicates are available, the bam files for the replicates can be provided as separate files, separated by whitespace. Bam files containing single-end reads and paired-end reads can be mixed since msCentipede currently does not model the fragment size distribution. However, bam files from different protocols or drastically different read lengths are best not mixed since protocol or read length differences could mask biologically meaningful heterogeneity that is relevant in identifying transcription factor binding sites. If the data were generated using an ATAC-seq protocol, the location of transpositions can be automatically identified from the read mapping positions by passing the flag <code>--protocol=ATAC_seq</code>.</p>

<h3>
<a id="learning-model-parameters" class="anchor" href="#learning-model-parameters" aria-hidden="true"><span class="octicon octicon-link"></span></a>Learning model parameters</h3>

<p>The model parameters can be learned by passing the following flags.</p>

<pre><code>python call_binding.py --task learn test/CTCF_chr10_motifs.txt.gz test/Gm12878_Rep1.bam test/Gm12878_Rep2.bam
</code></pre>

<p>This will run msCentipede with all other default values and output a log file <code>test/CTCF_chr10_motifs_msCentipede_log.txt</code> and a file <code>test/CTCF_chr10_motifs_msCentipede_model_parameters.pkl</code> in which the model parameter objects are stored. This is a standard Python pickle file that can be viewed using the <code>cPickle</code> module.</p>

<h3>
<a id="inferring-factor-binding" class="anchor" href="#inferring-factor-binding" aria-hidden="true"><span class="octicon octicon-link"></span></a>Inferring factor binding</h3>

<p>The posterior log odds of binding for a set of motif instances can be computed by passing the following flags.</p>

<pre><code>python call_binding.py --task infer test/CTCF_chr10_motifs.txt.gz test/Gm12878_Rep1.bam test/Gm12878_Rep2.bam
</code></pre>

<p>This will run msCentipede with all other default values and output a file <code>test/CTCF_chr10_motifs_msCentipede_binding_posterior.txt.gz</code>.</p>

<h3>
<a id="plotting-the-factor-specific-cleavage-profile" class="anchor" href="#plotting-the-factor-specific-cleavage-profile" aria-hidden="true"><span class="octicon octicon-link"></span></a>Plotting the factor-specific cleavage profile</h3>

<p>In order to check whether the algorithm could successfully learn a chromatin accessibility profile for a given set of motifs, we have provided a script to plot the shape of the inferred cleavage profile at factor bound sites.</p>

<pre><code>$ python plot_accessibility_profile.py

positional arguments:
  motif_file            name of a gzipped text file containing positional
                        information and other attributes for motif instances
                        of a transcription factor. columns of the file should
                        be as follows. Chromosome Start End Strand PWM_Score
                        [Attribute_1 Attribute_2 ...]. additional attributes
                        are optional.

optional arguments:
  --protocol {ATAC_seq,DNase_seq}
                        specifies the chromatin accessibility protocol
                        (default:DNase_seq)
  --model {msCentipede,msCentipede_flexbg,msCentipede_flexbgmean}
                        a different background cleavage profile will be shown
                        depending on the model (default:msCentipede)

</code></pre>

<h3>
<a id="optional-parameters" class="anchor" href="#optional-parameters" aria-hidden="true"><span class="octicon octicon-link"></span></a>Optional parameters</h3>

<p>Instead of the default file names, you can specify the file name to which the run log, model parameters and binding posterior odds will be written, using the flags <code>--log_file</code>, <code>--model_file</code> and <code>--posterior_file</code>, respectively. For example, these flags allow using a previously learned model on a new set of motif sites for the same transcription factors.</p>

<p>The differences between the three models <em>msCentipede</em> , <em>msCentipede_flexbgmean</em> , and <em>msCentipede_flexbg</em> are specified in detail in the associated <a href="">publication</a>. If the model flag is specified to be <em>msCentipede_flexbgmean</em> or <em>msCentipede_flexbg</em>, then a path to a bam file containing chromatin accessibility data from genomic DNA must be passed, using the flag <code>--bam_file_genomicdna</code>.</p>
      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">msCentipede maintained by <a href="https://github.com/rajanil">rajanil</a></p>
        <p>Published with <a href="https://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

              <script type="text/javascript">
            var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
            document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
          </script>
          <script type="text/javascript">
            try {
              var pageTracker = _gat._getTracker("UA-57610622-2");
            pageTracker._trackPageview();
            } catch(err) {}
          </script>


  </body>
</html>
